compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/FirSourceChild.kt

```diff
@@ -38,3 +40,39 @@ private fun FirLightSourceElement.getChild(types: Set<IElementType>, index: Int,
     val visitor = LighterTreeElementFinderByType(treeStructure, types, index, depth)
     return visitor.find(lighterASTNode)?.toFirLightSourceElement(treeStructure)
 }
+
+fun FirSourceElement.getParent(type: IElementType, strict: Boolean = false): FirSourceElement? {
+    return getParent(setOf(type), strict)
+}
+
+fun FirSourceElement.getParent(types: TokenSet, strict: Boolean = false): FirSourceElement? {
+    return getParent(types.types.toSet(), strict)
+}
+
+fun FirSourceElement.getParent(types: Set<IElementType>, strict: Boolean = false): FirSourceElement? {
+    return when (this) {
+        is FirPsiSourceElement<*> -> {
+            getParent(types, strict)
+        }
+        is FirLightSourceElement -> {
+            getParent(types, strict)
+        }
+        else -> null
+    }
+}
+
+private fun FirPsiSourceElement<*>.getParent(types: Set<IElementType>, strict: Boolean): FirSourceElement? {
+    var parent: PsiElement? = if (strict) psi.parent else psi
```

Comment:
The notion of `strict` looks confusing. What I thought was parent is actually strict parent, while what you can call parent by default is actually yourself. :\

How about naming like this: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/tree/src/org/jetbrains/kotlin/fir/scopes/FirScope.kt#L55

You can still put these utils, with a slight name change: `s/strict/includingSelf/g`; introduce two wrappers: `getParent` as above, and `getParentOrSelf`.

Reply:
OK, how about `getParent(types: ..., includingSelf: Boolean = false)`? This way I don't have to create 3 extra `getParentOrSelf` flavor of these.

Reply:
That sounds good.